File: top-interview-150/c++/530_minimum-absolute-difference-in-bst.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<int> tv{};
    int min_diff;

    void flat(TreeNode* root) {
        if (!root) return;
        if(root->val==(int)root->val) tv.push_back(root->val);
        flat(root->left);
        flat(root->right);
    }

    int getMinimumDifference(TreeNode* root) {
        flat(root);
        sort(tv.begin(), tv.end()); 
        //cout << "[";
        //for (auto v : tv)
        //    std::cout << v << " ";
        //cout << "]" << endl;
        int min_diff = tv[1]-tv[0];
        for (int i=1; i<tv.size()-1; i++) {
            min_diff = min(min_diff, tv[i+1] - tv[i]);
        }
        return min_diff;
    }
};

File: top-interview-150/python/530_minimum-absolute-difference-in-bst.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def flat(self, root):
        if root is None:
            return
        self.flat_list.append(root.val)
        self.flat(root.left)
        self.flat(root.right)

    def getMinimumDifference(self, root: Optional[TreeNode]) -> int:
        self.flat_list = []
        self.flat(root)
        self.flat_list = list(sorted([x for x in self.flat_list if x is not None]))
        #print(self.flat_list)
        self.min_diff = self.flat_list[1] - self.flat_list[0]
        for x in range(1, len(self.flat_list)-1):
            t = self.flat_list[x+1] - self.flat_list[x]
            #print(self.flat_list[x+1], self.flat_list[x], t)
            if t < self.min_diff:
                self.min_diff = t
        return self.min_diff
        